src/rust/catnap/futures/pop.rs

```diff
@@ -69,18 +76,29 @@ impl PopFuture {
 
 /// Future Trait Implementation for Pop Operation Descriptors
 impl Future for PopFuture {
-    type Output = Result<Buffer, Fail>;
+    type Output = Result<(Option<SocketAddrV4>, Buffer), Fail>;
 
     /// Polls the target [PopFuture].
     fn poll(self: Pin<&mut Self>, ctx: &mut Context<'_>) -> Poll<Self::Output> {
         let self_: &mut PopFuture = self.get_mut();
         let mut bytes: [u8; POP_SIZE] = [0; POP_SIZE];
-        match socket::recv(self_.fd, &mut bytes[..], socket::MsgFlags::empty()) {
+        match socket::recvfrom::<SockaddrStorage>(self_.fd, &mut bytes[..]) {
             // Operation completed.
-            Ok(nbytes) => {
+            Ok((nbytes, socketaddr)) => {
                 trace!("data received ({:?}/{:?} bytes)", nbytes, POP_SIZE);
                 let buf: Buffer = Buffer::Heap(DataBuffer::from_slice(&bytes[0..nbytes]));
-                Poll::Ready(Ok(buf))
+                let addr: Option<SocketAddrV4> = match socketaddr {
+                    Some(addr) => match addr.as_sockaddr_in() {
+                        Some(sin) => {
+                            let ip: Ipv4Addr = Ipv4Addr::from(sin.ip());
+                            let port: u16 = sin.port();
+                            Some(SocketAddrV4::new(ip, port))
```

Comment:
Don't you need to perform network-to-host byte-order conversion here?  nix::sys::socket::recvfrom looks to just be returning the sockaddr in a byte array, which would presumably yield big-endian values.  Unless something funky is happening in "as_sockaddr_in" (which I can't seem to find documentation for, where is that coming from?).

Reply:
@BrianZill, we don't need to because `as_sockaddr_in().ip()` and `as_sockaddr_in().port()` return the IP address and port number in host endianness, respectively. See: https://docs.rs/nix/latest/nix/sys/socket/struct.SockaddrIn.html

Reply:
Cool, thanks.